### PR TITLE
Add a daemon flag to prevent contact with v1 registries.

### PR DESCRIPTION
**Pull request by RichardScothern**
```
Prevent push and pull to v1 registries by filtering the available endpoints.
Add a daemon flag to control this behaviour.  Add a warning message when pulling
an image from a v1 registry.  The default order of pull is slightly altered
with this changset.

Previously it was:
https v2, https v1, http v2, http v1

now it is:
https v2, http v2, https v1, http v1

Prevent login to v1 registries by explicitly setting the version before ping to
prevent fallback to v1.

Add unit tests for v2 only mode.  Create a mock server that can register
handlers for various endpoints.  Assert no v1 endpoints are hit with legacy
registries disabled for the following commands:  pull, push, build, run and
login.  Assert the opposite when legacy registries are not disabled.
```

### CODE REVIEW

**Comment by tonistiigi** [comment]
Search does not seem to work for me with this option on. Getting `Error response from daemon: invalid registry endpoint https://index.docker.io/v2/`

**Comment by RichardScothern** [comment]
@tonistiigi : the only search endpoint in the hub is v1 (as the v2 registry does not support search).  

There are a few ways around this; probably redirecting or rewriting the URL in the hub instance.

**Comment by LK4D4** [comment]
I put `design-review` until we decide what to do with search.

File: integration-cli/docker_cli_v2_only.go [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+		err := os.Remove(f.Name())
+		if err != nil {
+			fmt.Println("Error removing tmpfile")
+		}
+	}
+	return f.Name(), cleanup, nil
+
+}
+
+// TestV2Only ensures that a deamon in v2-only mode does not
```
Comment (aaronlehmann) [inline integration-cli/docker_cli_v2_only.go]: s/deamon/daemon/

**Comment by aaronlehmann** [comment]
The code looks good to me. I think it's reasonable for this option to break search, since it's a v1-only feature. If we want to keep search working, we could change this option from "disallow communication with v1 registries" to "disallow pushes and pulls to/from v1 registries".

**Comment by thaJeztah** [comment]
I'm assuming this will be the default in the future? If so, we might consider renaming this
option to, e.g. `--enable-legacy-registry`, and setting it to `true` by default for now. To *disable* legacy registries, users can set `--enable-legacy-registry=false`.

Once we default to disabling registries, the default value will be `false`, and users can add the `--enable-legacy` to explicitly allow legacy registries.

If changing the default is already on the roadmap, I think we should add a deprecation message to [docs/misc/deprecated.md](https://github.com/docker/docker/blob/master/docs/misc/deprecated.md), including the version in which the default will change, so that users can prepare for this change in behavior (if needed).

**Comment by RichardScothern** [comment]
Search has been reenabled (though no verbiage has changed).

A warning message has been added on v1 pull.  It reads
`localhost:5000/hello-world: this image was pulled from a legacy registry.  Important: These registries will not be supported in future versions of docker.`

**Comment by tiborvass** [comment]
I agree with keeping search working and with the warning message.

> If we want to keep search working, we could change this option from "disallow communication with v1 registries" to "disallow pushes and pulls to/from v1 registries".

But I also agree with @aaronlehmann's statement. So `--enable-legacy-registry` is too vague, since it only refers to push/pull and login but not search. Not sure what a better flag name would be

Also what about search for private images?

Ping @RichardScothern @thaJeztah

File: integration-cli/docker_cli_v2_only.go [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+		w.WriteHeader(404)
+	})
+
+	reg.registerHandler("/v1/.*", func(w http.ResponseWriter, r *http.Request) {
+		c.Fatal("V1 registry contacted")
+	})
+
+	repoName := fmt.Sprintf("%s/busybox", reg.hostport)
+
+	err = s.d.Start(fmt.Sprintf("--insecure-registry=%s", reg.hostport), "--disable-legacy-registry=true")
```
Comment (calavera) [inline integration-cli/docker_cli_v2_only.go]: you can do this without that Sprintf:

```go
s.d.Start("--insecure-registry", reg.hostport, "--disable-legacy-registry")
```

File: registry/config.go [hunk 3]
```diff
@@ -20,6 +20,10 @@ type Options struct {
 	InsecureRegistries opts.ListOpts
 }
 
+// V2Only controls access to legacy registries.  If it is set to true via the
+// command line flag the daemon will not attempt to contact legacy registries
+var V2Only bool
```
Comment (calavera) [inline registry/config.go]: can you put this var in the `var` block a few lines below?

File: registry/service.go [hunk 4]
```diff
@@ -40,7 +36,14 @@ func (s *Service) Auth(authConfig *cliconfig.AuthConfig) (string, error) {
 	if err != nil {
 		return "", err
 	}
-	endpoint, err := NewEndpoint(index, nil)
+
+	endpointVersion := APIVersion(APIVersionUnknown)
```
Comment (calavera) [inline registry/service.go]: I don't think this cast is necessary, is it? ApiVersionUnknown is already an APIVersion. Same in line 63.

File: registry/service_v1.go [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+	if slashIndex <= 0 {
+		return nil, fmt.Errorf("invalid repo name: missing '/':  %s", repoName)
+	}
+	hostname := repoName[:slashIndex]
+
+	tlsConfig, err = s.TLSConfig(hostname)
+	if err != nil {
+		return nil, err
+	}
+	isSecure := !tlsConfig.InsecureSkipVerify
```
Comment (calavera) [inline registry/service_v1.go]: this variable doesn't need to be defined. You can do `if !tlsConfig.InsecureSkipVerify` in like 44.

File: integration-cli/docker_cli_v2_only.go [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+	v1Repo := 0
+	reg.registerHandler("/v1/repositories/busybox/", func(w http.ResponseWriter, r *http.Request) {
+		v1Repo++
+	})
+
+	reg.registerHandler("/v1/repositories/busybox/images", func(w http.ResponseWriter, r *http.Request) {
+		v1Repo++
+	})
+
+	err = s.d.Start(fmt.Sprintf("--insecure-registry=%s", reg.hostport), "--disable-legacy-registry=false")
```
Comment (calavera) [inline integration-cli/docker_cli_v2_only.go]: the Sprintf is not necessary here either.

**Comment by tiborvass** [comment]
@RichardScothern FYI your 2 tests are failing because `flag provided but not defined: --disable-legacy-registry`.
You can see that in `./bundles/latest/test-integration-cli/dXXXXXXX/docker.log` where `dXXXXXX` is the daemon instance's name as shown in the error output.

File: registry/service.go (hunk 4 above)
Comment (RichardScothern) [inline registry/service.go]: APIVersion is a type.  APIVersionUnknown is an int (generated by iota), hence the cast.

**Comment by tiborvass** [comment]
What do you guys think of `--legacy-registry=pull,push,search,login` describing what's currently the case, and `--legacy-registry=search` what would now be the default.

Also, does this mean it breaks private search by default? Also, shouldn't this be a flag on the client instead of the daemon? `docker pull --legacy-mode` etc. ?

**Comment by RichardScothern** [comment]
I don't see a reason to offer such a fine-grained control options when most combinations don't make sense.   I think better naming & description of  the flag is the way forward.

I don't see a problem with private registry search.  On a private v2 registry search is not implemented (no behavior change).  On a v1/v2 private registry, search will fallback to v1 with the flag set to either value (no behavior change).  On a private v1 registry, it makes no sense to set the flag to true as nothing will work.

The flag is on the daemon with other registry specific flags (e.g. `--insecure-registry`).

**Comment by icecrime** [comment]
I don't think we need fined grained control: this only exists to prepare users to the full deprecation of the v1 protocol on the Engine side, which we can safely assume will happen before the end of the year. The flag should allow to answer the question: "I am ready for a pure v2 world".

For the same reason, I believe `--no-legacy-registry` is just fine as the flag itself will go away with the v1 protocol code (or maybe we'll just ignore its value).

Regarding `search`, I'm slightly confused by the conversation. My idea is that search is a v1 feature that you accept losing if you deny communication with v1 registries. I don't expect any special case here.

**Comment by icecrime** [comment]
Let me rephrase that: it's unfortunate that search is a v1 thing and that the feature gets lost in the process. However I do believe that this is a different problem (which should most likely be addressed with v2 supporting search).

File: integration-cli/docker_cli_v2_only.go [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+	}
+	defer cleanup()
+
+	s.d.Cmd("build", "--file", dockerfileName, ".")
+
+	s.d.Cmd("run", repoName)
+	s.d.Cmd("login", "-u", "richard", "-p", "testtest", "-e", "testuser@testdomain.com", reg.hostport)
+	s.d.Cmd("tag", "busybox", repoName)
+	s.d.Cmd("push", repoName)
+	s.d.Cmd("pull", repoName)
```
Comment (tiborvass) [inline integration-cli/docker_cli_v2_only.go]: @RichardScothern dont' you need to check for errors returned by Cmd ?

File: integration-cli/registry_mock.go [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+
+type handlerFunc func(w http.ResponseWriter, r *http.Request)
+
+type testRegistry struct {
+	server   *httptest.Server
+	hostport string
+	handlers map[string]handlerFunc
+}
+
+func (tr testRegistry) registerHandler(path string, h handlerFunc) {
```
Comment (tiborvass) [inline integration-cli/registry_mock.go]: add a comment saying that this can't be called concurrently

File: registry/config.go [hunk 9]
```diff
@@ -44,6 +44,10 @@ var (
 	ErrInvalidRepositoryName = errors.New("Invalid repository name (ex: \"registry.domain.tld/myrepos\")")
 
 	emptyServiceConfig = NewServiceConfig(nil)
+
+	// V2Only controls access to legacy registries.  If it is set to true via the
+	// command line flag the daemon will not attempt to contact legacy registries
```
Comment (tiborvass) [inline registry/config.go]: you can be more precise and say that `legacy (V1) registries`

File: registry/config.go [hunk 10]
```diff
@@ -53,6 +57,7 @@ func (options *Options) InstallFlags(cmd *flag.FlagSet, usageFn func(string) str
 	cmd.Var(&options.Mirrors, []string{"-registry-mirror"}, usageFn("Preferred Docker registry mirror"))
 	options.InsecureRegistries = opts.NewListOpts(ValidateIndexName)
 	cmd.Var(&options.InsecureRegistries, []string{"-insecure-registry"}, usageFn("Enable insecure registry communication"))
+	cmd.BoolVar(&V2Only, []string{"-no-legacy-registry"}, false, "Don't contact legacy registries")
```
Comment (tiborvass) [inline registry/config.go]: `Do not`

File: registry/service.go (hunk 4 above)
Comment (tiborvass) [inline registry/service.go]: We should fix that in a separate PR

File: registry/config.go [hunk 11]
```diff
@@ -44,6 +44,10 @@ var (
 	ErrInvalidRepositoryName = errors.New("Invalid repository name (ex: \"registry.domain.tld/myrepos\")")
 
 	emptyServiceConfig = NewServiceConfig(nil)
+
+	// V2Only controls access to legacy registries.  If it is set to true via the
+	// command line flag the daemon will not attempt to contact legacy registries
+	V2Only = false
```
Comment (tiborvass) [inline registry/config.go]: how about we make this `runtime.GOOS != "linux"` that way new OSes automatically won't inherit v1.

File: registry/service_v2.go [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+			Version:      APIVersion2,
+			Official:     true,
+			TrimHostname: true,
+			TLSConfig:    tlsConfig,
+		})
+
+		return endpoints, nil
+	}
+
+	slashIndex := strings.IndexRune(repoName, '/')
```
Comment (tiborvass) [inline registry/service_v2.go]: this is duplicate code from the v1 lookup too. you could have a hlper function that takes repoName and returns a tlsConfig and an error.

**Comment by tiborvass** [comment]
@RichardScothern It seems like there is one big change in this PR that needs to be highlighted.

Where previously the prioritization was: v2 mirrors (for hub images), HTTPS v2 registry, HTTPS v1 registry, HTTP v2 registry, HTTP v1 registry,
it is now: v2 mirrors (for hub images), HTTPS v2 registry, HTTP v2 registry, HTTPS v1 registry, HTTP v1 registry.

The change is basically about making HTTP v2 be considered better than HTTPS v1.

If we accept this, then I think this has to be stated in the description of this PR, in the commit, and possibly in a comment on `lookupEndpoints`.

/cc @dmcgowan @dmp42

**Comment by dmcgowan** [comment]
@tiborvass I discussed this with @RichardScothern and @dmp42 earlier and we are in agreement that this change in behavior is acceptable. There is no valid case we could think of where this ordering will cause issues since it would already require setting a registry as insecure, a setting which should not be used when MITM is a possibility. If you can think of a case where you think this is not acceptable do raise the issue.

I don't think a comment is necessary in `lookupEndpoints` since it is already cleanly breaking up the v2 and v1 code paths, so it should not be able to cause any confusion. I think it is fine to call it out in the commit since this is a behavior change.

**Comment by tiborvass** [comment]
@dmcgowan thanks for the answer, you got a point about my list being under the assumption the registry was marked insecure. So it is indeed perfectly fine.

Agreed for not needing a comment in the code, but needed in the commit message and description of PR.

Thanks very much!

File: registry/service_v2.go (hunk 12 above)
Comment (RichardScothern) [inline registry/service_v2.go]: It would need to return the hostname too.  I don't think it will save too much code and the v1 version will be removed.

File: registry/config.go (hunk 11 above)
Comment (RichardScothern) [inline registry/config.go]: done in init()

File: integration-cli/docker_cli_v2_only.go (hunk 7 above)
Comment (RichardScothern) [inline integration-cli/docker_cli_v2_only.go]: Strictly no.  If any commands fail the expected endpoint hits will not be as expected and the tests will fail anyway.

Comment (tiborvass) [inline integration-cli/docker_cli_v2_only.go]: Fair enough

**Comment by icecrime** [comment]
Tested in simple cases, and verified that `--insecure-registry` is still honored (both locally and remotely). LGTM

**Comment by tiborvass** [comment]
LGTM the windows test failures are unrelated

Tested manually, and with derek's integration tests to make sure tls and auth are still working as expected.

**Comment by tiborvass** [comment]
@RichardScothern please update the documentation

**Comment by RichardScothern** [comment]
Documentation added.

**Comment by icecrime** [comment]
We will need to add doc for this in the (not yet existing) `docker-daemon` manual (see #16063).

**Comment by RichardScothern** [comment]
@icecrime: new manual is updated.

**Comment by thaJeztah** [comment]
docs LGTM, thanks

**Comment by RichardScothern** [comment]
I think this is good to merge.  The Windows build failure is not code-related.

**Comment by jfrazelle** [comment]
wtf I hate the flag name
